src/libcore/pin.rs

```diff
@@ -94,6 +95,106 @@
 //! // let new_unmoved = Unmovable::new("world".to_string());
 //! // std::mem::swap(&mut *still_unmoved, &mut *new_unmoved);
 //! ```
+//!
+//! # `Drop` guarantee
+//!
+//! The purpose of pinning is to be able to rely on the placement of some data in memory.
+//! To make this work, not just moving the data is restricted; deallocating or overwriting
+//! it is restricted, too. Concretely, for pinned data you have to maintain the invariant
+//! that *it will not get overwritten or deallocated until `drop` was called*.
+//! ("Overwriting" here refers to other ways of invalidating storage, such as switching
+//! from one enum variant to another.)
+//!
+//! The purpose of this guarantee is to allow data structures that store pointers
+//! to pinned data. For example, in an intrusive doubly-linked list, every element
+//! will have pointers to its predecessor and successor in the list. Every element
+//! will be pinned, because moving the elements around would invalidate the pointers.
```

Comment:
This is a good example that I think should come further up instead of being buried in the `Drop` section. And then we can refer back to it here.

Reply:
Well but then I have to write actual code that, you know, compiles. Writing text is so much faster...^^

Reply:
True, but I think that's worthwhile still :p And also what @Centril is asking for above.

Reply:
I think I'll make this a text-only example. I'm afraid before the week-end I won't have the time to full in code for this. Just applying all that feedback will already take a while.^^

But if someone wants to help, that'd be great :D

Reply:
Resolved. The textual example helps.